pilot/pkg/model/sidecar.go

```diff
@@ -908,3 +842,66 @@ func needsPortMatch(l *networking.IstioEgressListener) bool {
 	return l != nil && l.Port.GetNumber() != 0 &&
 		protocol.Parse(l.Port.Protocol) != protocol.HTTP_PROXY
 }
+
+type sidecarServiceIndex struct {
+	svc   *Service
+	index int // index record the position of the svc in slice
+}
+
+func (sc *SidecarScope) appendSidecarServices(servicesAdded map[host.Name]sidecarServiceIndex, s *Service) {
+	if s == nil {
+		return
+	}
+	if foundSvc, found := servicesAdded[s.Hostname]; !found {
+		sc.services = append(sc.services, s)
+		servicesAdded[s.Hostname] = sidecarServiceIndex{s, len(sc.services) - 1}
+		sc.servicesByHostname[s.Hostname] = s
+	} else {
+		existing := foundSvc.svc
+		// We donot merge k8s service with any other services from other registries
+		if existing.Attributes.ServiceRegistry == provider.Kubernetes {
+			return
+		}
+		// In some scenarios, there may be multiple Services defined for the same hostname due to ServiceEntry allowing
+		// arbitrary hostnames. In these cases, we want to pick the first Service, which is the oldest. This ensures
+		// newly created Services cannot take ownership unexpectedly.
+		// However, the Service is from Kubernetes it should take precedence over ones not. This prevents someone from
+		// "domain squatting" on the hostname before a Kubernetes Service is created.
+		if s.Attributes.ServiceRegistry == provider.Kubernetes {
+			log.Debugf("Service %s/%s from registry %s ignored by %s/%s/%s", existing.Attributes.Namespace, existing.Hostname, existing.Attributes.ServiceRegistry,
```

Comment:
This debug log line is incorrect now?

Reply:
what do you mean

Reply:
nvm.. we come to this only if `existing` is service entry